rust/flatbuffers/src/builder.rs

```diff
@@ -62,24 +62,32 @@ pub struct FlatBufferBuilder<'fbb> {
 impl<'fbb> FlatBufferBuilder<'fbb> {
     /// Create a FlatBufferBuilder that is ready for writing.
     pub fn new() -> Self {
-        Self::new_with_capacity(0)
+        Self::with_capacity(0)
+    }
+    #[deprecated(note = "replaced with `with_capacity`", since = "0.8.5")]
+    pub fn new_with_capacity(size: usize) -> Self {
+        Self::with_capacity(size)
     }
-
     /// Create a FlatBufferBuilder that is ready for writing, with a
     /// ready-to-use capacity of the provided size.
     ///
     /// The maximum valid value is `FLATBUFFERS_MAX_BUFFER_SIZE`.
-    pub fn new_with_capacity(size: usize) -> Self {
+    pub fn with_capacity(size: usize) -> Self {
+        Self::from_vec(vec![0; size])
+    }
+    /// Create a FlatBufferBuilder that is ready for writing, reusing
+    /// an existing vector.
+    pub fn from_vec(buffer: Vec<u8>) -> Self {
```

Comment:
Should we have an optional "starting" location of the buffer? The whole point of the issue was to prefix the 16 bytes, but currently nothing prevents those spots from used in the normal course of filling out the buffer.

Reply:
If the "starting location" is `head`, that would create a suffix, not prefix, since we build Flatbuffers backwards.

I think you are asking  when `collapse -> (buffer, head)` is called,
we guarantee `buffer.len() - head >= minimum_prefix_size`.
Basically we would need `fn ensure_capacity(&mut self, want: usize) -> usize;`
to always reserve `want + minimum_prefix_size` capacity.

This is doable, but perhaps too niche imo. Am I understanding correctly?

Reply:
Hmm I think it's too niche--the user can always appropriately size the buffer before passing in.
However... looking at the use-case... calling collapse then from_vec leaves the following variables with *new* memory
```
            field_locs: Vec::new(),
            written_vtable_revpos: Vec::new(),
            strings_pool: Vec::new(),
```
As a user of the library I'd rather do the extra memcpy or the unsupported backwards push than "lose" the allocation of those critical internal resources and have them be reallocated for every packet.  Would it be possible to make a mutable "view" of the internal buffer vec (if the finished flag is set) instead of asking users to collapse and reconstruct the whole class?

Reply:
Hmm, I was more thinking of reserving space at the front of the buffer (i guess we call it the tail since we build backwards) but I guess that won't always work, since the tail might not butt up against the head location. So forget my suggestion.

So your suggestion of `minimum_prefix_size` would be better, so any resizing would automatically include enough space. 

I think for the desired application, it reduces the need to do a full memcpy of the finalized buffer into another buffer just to prepend data. So while niche, it is an important optimization if it's in the network stack layer.

Reply:
@danielrh, I think you mean a method like:
```rust
/// Returns a mutable view of the internal buffer and location of where the flatbuffer data starts.
/// Note that modifying the flatbuffer data may corrupt it. 
/// Panics if the buffer has not been finished.
fn mut_finished_buffer(&mut self) -> (&mut [u8], usize);
```
WDYT?